### PR TITLE
Add loading yaml label template files

### PR DESCRIPTION
**Pull request by lafriks**
Extract from #11669 and enhancement to #22585 to support exclusive scoped labels in label templates

* Move label template functionality to label module
* Fix handling of color codes
* Add Advanced label template

### CODE REVIEW

**Review by lunny — APPROVED** [review]

File: services/migrations/gitea_uploader.go [hunk 1]
```diff
@@ -217,18 +218,19 @@ func (g *GiteaLocalUploader) CreateMilestones(milestones ...*base.Milestone) err
...
 
 		lbs = append(lbs, &issues_model.Label{
 			RepoID:      g.repo.ID,
-			Name:        label.Name,
-			Description: label.Description,
-			Color:       "#" + label.Color,
+			Name:        l.Name,
+			Exclusive:   l.Exclusive,
+			Description: l.Description,
+			Color:       "#" + l.Color,
```
Comment (wxiaoguang) [inline services/migrations/gitea_uploader.go]: `#RGB` will pass the ColorPattern check, and be used as `label.Color`, but it's not a valid color (not normalized), will break functions like ColorRGB/UseLightTextColor, etc.


IMO, the `ColorPattern` should be removed, only use `color, err := NormalizeColor`

Comment (lafriks) [reply services/migrations/gitea_uploader.go]: @wxiaoguang done

File: modules/label/label.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+package label
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+// colorPattern is a regexp witch can validate label color
```
Comment (delvh) [inline modules/label/label.go]:
```suggestion
// colorPattern is a regexp which can validate label color
```

File: services/migrations/gitea_uploader.go [hunk 3]
```diff
@@ -217,18 +218,20 @@ func (g *GiteaLocalUploader) CreateMilestones(milestones ...*base.Milestone) err
...
 	lbs := make([]*issues_model.Label, 0, len(labels))
-	for _, label := range labels {
-		// We must validate color here:
-		if !issues_model.LabelColorPattern.MatchString("#" + label.Color) {
-			log.Warn("Invalid label color: #%s for label: %s in migration to %s/%s", label.Color, label.Name, g.repoOwner, g.repoName)
-			label.Color = "ffffff"
+	for _, l := range labels {
+		if color, err := label.NormalizeColor(l.Color); err != nil {
+			log.Warn("Invalid label color: #%s for label: %s in migration to %s/%s", l.Color, l.Name, g.repoOwner, g.repoName)
+			l.Color = "#ffffff"
```
Comment (delvh) [inline services/migrations/gitea_uploader.go:224]: Won't the leading `#` now be somewhat breaking compared to earlier where it wasn't present?

File: modules/label/parser.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+// IsErrTemplateLoad checks if an error is a ErrTemplateLoad.
+func IsErrTemplateLoad(err error) bool {
+	_, ok := err.(ErrTemplateLoad)
+	return ok
+}
+
+func (err ErrTemplateLoad) Error() string {
+	return fmt.Sprintf("Failed to load label template file '%s': %v", err.TemplateFile, err.OriginalError)
```
Comment (delvh) [inline modules/label/parser.go:33]:
```suggestion
	return fmt.Sprintf("Failed to load label template file '%s': %w", err.TemplateFile, err.OriginalError)
```

File: modules/label/parser.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+	if err := yaml.Unmarshal(data, lf); err != nil {
+		return nil, err
+	}
+
+	// Validate label data and fix colors
+	for _, l := range lf.Labels {
+		l.Color = strings.TrimSpace(l.Color)
+		if len(l.Name) == 0 || len(l.Color) == 0 {
+			return nil, ErrTemplateLoad{name, errors.New("label name and color are required fields")}
```
Comment (delvh) [inline modules/label/parser.go:68]: Color is required?
Shouldn't we just assign a random color when it isn't set?
(But that's an improvement for the future, not necessarily for this PR)

File: modules/label/parser.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+		}
+		color, err := NormalizeColor(l.Color)
+		if err != nil {
+			return nil, ErrTemplateLoad{name, fmt.Errorf("bad HTML color code in label: %s", l.Name)}
```
Comment (delvh) [inline modules/label/parser.go]: `l.Name`?
```suggestion
			return nil, ErrTemplateLoad{name, fmt.Errorf("bad HTML color code in label: %s", l.Color)}
```
or
```suggestion
			return nil, ErrTemplateLoad{name, fmt.Errorf("bad HTML color code in label %s: %s", l.Name, l.Color)}
```

File: modules/label/parser.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+		}
+		l.Color = color
+	}
+
+	return lf.Labels, nil
+}
+
+func parseDefaultFormat(name string, data []byte) ([]*Label, error) {
```
Comment (delvh) [inline modules/label/parser.go]: Isn't that `parseLegacyFormat` now?

File: modules/label/parser_test.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+	assert.True(t, labels[2].Exclusive)
+	assert.Equal(t, "#ee0000", labels[2].Color)
+	assert.Equal(t, "High priority", labels[2].Description)
+	assert.Equal(t, "type/bug", labels[3].Name)
+	assert.False(t, labels[3].Exclusive)
+	assert.Equal(t, "#ff0000", labels[3].Color)
+	assert.Equal(t, "Bug", labels[3].Description)
+}
+
+func TestDefaultParser(t *testing.T) {
```
Comment (delvh) [inline modules/label/parser_test.go]:
```suggestion
func TestLegacyParser(t *testing.T) {
```

File: modules/label/parser.go (hunk 4 above)
Comment (lafriks) [reply modules/label/parser.go:33]: There is is no point in this as this returns string not error and also imho %w is only supported in `fmt.Errorf`

File: modules/label/parser.go (hunk 5 above)
Comment (lafriks) [reply modules/label/parser.go:68]: It is and was also previously

File: modules/label/parser.go (hunk 4 above)
Comment (delvh) [reply modules/label/parser.go:33]: Yep, I confused it with `fmt.Errorf` as it looked like you would return an error.

File: services/migrations/gitea_uploader.go (hunk 3 above)
Comment (lafriks) [reply services/migrations/gitea_uploader.go:224]: It always was only it was adding it as prefix later in code, now that part has been moved into NormalizeColor so it's needed to be added here

File: modules/label/parser.go (hunk 4 above)
Comment (lafriks) [reply modules/label/parser.go:33]: `fmt.Sprintf does not support error-wrapping directive %w`

**Comment by lafriks** [comment]
@delvh updated code as per suggestions

**Review by delvh — APPROVED** [review]

**Pull request merged**